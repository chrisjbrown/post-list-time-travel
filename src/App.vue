<template>
  <div
    class="absolute -z-10 border-b-[10vw] border-l-[100vw] border-b-transparent border-l-indigo-900"
  ></div>
  <div class="mx-auto grid max-w-[1200px] gap-12 p-8 grid-cols-[1fr_1fr]">
    <PostList :posts="posts" @action="onMove" />
    <ActionList :actions="actions" @travel="timeTravel" />
  </div>
</template>

<script setup lang="ts">
import PostList from "./components/PostList.vue";
import ActionList from "./components/ActionList.vue";

import { ref } from "vue";
import type { PostType, ActionType } from "./components/types";
const posts = ref<PostType[]>([]);
const actions = ref<ActionType[]>([]);

async function getPosts() {
  // could add more error handling here
  const response = await fetch("https://jsonplaceholder.typicode.com/posts");
  const allPosts = (await response.json()) as unknown;

  if (!Array.isArray(allPosts)) throw new Error("Array expected");

  posts.value = allPosts.slice(0, 5).map((post) => {
    return {
      id: post.id,
    };
  });
}
getPosts();

/**
 * Move post and update actions
 * @param action ActionType
 */
function onMove(action: ActionType) {
  movePost(action.from, action.to);
  actions.value = [action, ...actions.value];
}

/**
 * Move post from to in actions array
 * @param from number - index of post to move
 * @param to number - index of where to move post to
 */
function movePost(from: number, to: number) {
  const postToMove = posts.value[from];
  posts.value.splice(from, 1);
  posts.value.splice(to, 0, postToMove);
}

/**
 * Move backwards through time based on number of jumps
 * @param jumps number of jumps to make
 */
function timeTravel(jumps: number) {
  for (let i = 0; i <= jumps; i++) {
    const nextAction = actions.value.shift();
    if (!nextAction) break;

    movePost(nextAction.to, nextAction.from);
  }
}
</script>
