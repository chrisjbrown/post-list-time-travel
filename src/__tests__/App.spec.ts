import { describe, expect, it, vi } from "vitest";
import { mount, VueWrapper, DOMWrapper } from "@vue/test-utils";

import App from "../App.vue";

const posts = [...Array(5).keys()].map((id) => ({ id: id + 1 }));

function getPosts(wrapper: VueWrapper): string[] {
  return wrapper
    .findAll('[data-testid^="post-"]')
    .map((post) => post.attributes("data-testid") || "");
}
function getActions(wrapper: VueWrapper): DOMWrapper<HTMLElement>[] {
  return wrapper.findAll('[data-testid^="action-"]');
}

describe("App", () => {
  const mockFetch = vi.spyOn(global, "fetch");
  mockFetch.mockReturnValue({
    json: () => new Promise((resolve) => resolve(posts)),
  });
  it("Consumes usePosts and renders", async () => {
    const wrapper = mount(App);

    await new Promise(process.nextTick);
    expect(wrapper.html()).toMatchSnapshot();
  });
  describe("Moves posts", async () => {
    const wrapper = mount(App);

    await new Promise(process.nextTick);
    await wrapper.find('[data-testid="icon-down"]').trigger("click");
    await wrapper.findAll('[data-testid="icon-up"]')?.[3].trigger("click");
    await wrapper.findAll('[data-testid="icon-up"]')?.[2].trigger("click");

    it("Order of posts are correct", async () => {
      expect(getPosts(wrapper)).toEqual([
        "post-2",
        "post-1",
        "post-5",
        "post-3",
        "post-4",
      ]);

      const actionList = wrapper.findAll('[data-testid^="action-"]');
      expect(actionList.length).toEqual(3);
    });
    it("Number of actions is correct", async () => {
      expect(getActions(wrapper).length).toEqual(3);
    });
    it("Time travel back one, results in correct order and actions", async () => {
      const timeTravelBtns = wrapper.findAll('[data-testid^="time-travel-"]');
      await timeTravelBtns[0].trigger("click");
      expect(getPosts(wrapper)).toEqual([
        "post-2",
        "post-1",
        "post-3",
        "post-5",
        "post-4",
      ]);

      expect(getActions(wrapper).length).toEqual(2);
    });
    it("Time travel back two, results in original order and no actions", async () => {
      const timeTravelBtns = wrapper.findAll('[data-testid^="time-travel-"]');
      await timeTravelBtns[1].trigger("click");
      expect(getPosts(wrapper)).toEqual([
        "post-1",
        "post-2",
        "post-3",
        "post-4",
        "post-5",
      ]);

      expect(getActions(wrapper).length).toEqual(0);
    });
  });
});
