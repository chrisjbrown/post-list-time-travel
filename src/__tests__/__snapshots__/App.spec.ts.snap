// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > Consumes usePosts and renders 1`] = `
"<div class="absolute -z-10 border-b-[10vw] border-l-[100vw] border-b-transparent border-l-indigo-900"></div>
<!--v-if-->
<div class="mx-auto grid max-w-[1200px] gap-12 p-8 grid-cols-[1fr] md:grid-cols-[1fr_1fr]">
  <div data-v-46953f88="">
    <h1 data-v-46953f88="" class="text-3xl font-bold text-white">Sortable List</h1>
    <transition-group-stub data-v-46953f88="" name="list" tag="ul" appear="false" persisted="false" css="true">
      <li data-v-46953f88="" data-testid="post-1">
        <div data-v-46953f88="" class="flex h-16 p-4 bg-white rounded-md my-4 items-center shadow-lg">
          <div data-v-46953f88="">Post 1</div>
          <div data-v-46953f88="" class="ml-3 flex flex-col">
            <!--v-if--><button data-v-46953f88="" data-testid="icon-down" aria-label="move down" class="hover:bg-slate-100"><svg data-v-46953f88="" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1" stroke="currentColor" class="w-5 h-5">
                <path stroke-linecap="round" stroke-linejoin="round" d="m19.5 8.25-7.5 7.5-7.5-7.5"></path>
              </svg></button>
          </div>
        </div>
      </li>
      <li data-v-46953f88="" data-testid="post-2">
        <div data-v-46953f88="" class="flex h-16 p-4 bg-white rounded-md my-4 items-center shadow-lg">
          <div data-v-46953f88="">Post 2</div>
          <div data-v-46953f88="" class="ml-3 flex flex-col"><button data-v-46953f88="" data-testid="icon-up" aria-label="move up" class="hover:bg-slate-100"><svg data-v-46953f88="" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1" stroke="currentColor" class="w-5 h-5">
                <path stroke-linecap="round" stroke-linejoin="round" d="m4.5 15.75 7.5-7.5 7.5 7.5"></path>
              </svg></button><button data-v-46953f88="" data-testid="icon-down" aria-label="move down" class="hover:bg-slate-100"><svg data-v-46953f88="" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1" stroke="currentColor" class="w-5 h-5">
                <path stroke-linecap="round" stroke-linejoin="round" d="m19.5 8.25-7.5 7.5-7.5-7.5"></path>
              </svg></button></div>
        </div>
      </li>
      <li data-v-46953f88="" data-testid="post-3">
        <div data-v-46953f88="" class="flex h-16 p-4 bg-white rounded-md my-4 items-center shadow-lg">
          <div data-v-46953f88="">Post 3</div>
          <div data-v-46953f88="" class="ml-3 flex flex-col"><button data-v-46953f88="" data-testid="icon-up" aria-label="move up" class="hover:bg-slate-100"><svg data-v-46953f88="" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1" stroke="currentColor" class="w-5 h-5">
                <path stroke-linecap="round" stroke-linejoin="round" d="m4.5 15.75 7.5-7.5 7.5 7.5"></path>
              </svg></button><button data-v-46953f88="" data-testid="icon-down" aria-label="move down" class="hover:bg-slate-100"><svg data-v-46953f88="" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1" stroke="currentColor" class="w-5 h-5">
                <path stroke-linecap="round" stroke-linejoin="round" d="m19.5 8.25-7.5 7.5-7.5-7.5"></path>
              </svg></button></div>
        </div>
      </li>
      <li data-v-46953f88="" data-testid="post-4">
        <div data-v-46953f88="" class="flex h-16 p-4 bg-white rounded-md my-4 items-center shadow-lg">
          <div data-v-46953f88="">Post 4</div>
          <div data-v-46953f88="" class="ml-3 flex flex-col"><button data-v-46953f88="" data-testid="icon-up" aria-label="move up" class="hover:bg-slate-100"><svg data-v-46953f88="" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1" stroke="currentColor" class="w-5 h-5">
                <path stroke-linecap="round" stroke-linejoin="round" d="m4.5 15.75 7.5-7.5 7.5 7.5"></path>
              </svg></button><button data-v-46953f88="" data-testid="icon-down" aria-label="move down" class="hover:bg-slate-100"><svg data-v-46953f88="" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1" stroke="currentColor" class="w-5 h-5">
                <path stroke-linecap="round" stroke-linejoin="round" d="m19.5 8.25-7.5 7.5-7.5-7.5"></path>
              </svg></button></div>
        </div>
      </li>
      <li data-v-46953f88="" data-testid="post-5">
        <div data-v-46953f88="" class="flex h-16 p-4 bg-white rounded-md my-4 items-center shadow-lg">
          <div data-v-46953f88="">Post 5</div>
          <div data-v-46953f88="" class="ml-3 flex flex-col"><button data-v-46953f88="" data-testid="icon-up" aria-label="move up" class="hover:bg-slate-100"><svg data-v-46953f88="" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1" stroke="currentColor" class="w-5 h-5">
                <path stroke-linecap="round" stroke-linejoin="round" d="m4.5 15.75 7.5-7.5 7.5 7.5"></path>
              </svg></button>
            <!--v-if-->
          </div>
        </div>
      </li>
    </transition-group-stub>
  </div>
  <div data-v-a1dcdb9c="" class="bg-white border flex-col flex rounded-md shadow-lg">
    <h1 data-v-a1dcdb9c="" class="text-3xl font-bold p-4">List of actions committed</h1>
    <div data-v-a1dcdb9c="" class="p-6 bg-slate-100 grow">
      <transition-group-stub data-v-a1dcdb9c="" name="fade" tag="ul" appear="false" persisted="false" css="true" class="bg-white rounded-md"></transition-group-stub>
    </div>
  </div>
</div>"
`;
